include/sound/soc.h

```diff
@@ -1106,7 +1106,7 @@ struct snd_soc_pcm_runtime {
 #define asoc_rtd_to_cpu(rtd, n)   (rtd)->dais[n]
 #define asoc_rtd_to_codec(rtd, n) (rtd)->dais[n + (rtd)->dai_link->num_cpus]
 #define asoc_substream_to_rtd(substream) \
-	(struct snd_soc_pcm_runtime *)snd_pcm_substream_chip(substream)
+	(struct snd_soc_pcm_runtime *)(substream)->pcm->private_data
```

Comment:
it's really unclear to me from the commit message what the benefits of moving the rtd storage from

substream->private_data to substream->pcm->private_data might be?

You are referring to a very obscure 'reparenting' DPCM capability but for the rest of us what does this mean?

Reply:
Hmm.. I tried to be as transparent in the cover letter as possible. The case is not a common one, indeed, but has 100% reproduction rate, regardless of the HDAudio driver chosen, be it avs or sof.

The reparenting functionality is what allows ASoC to handle Nx FE <-> 1x BE and so on. Let's assume we are dealing with: 2x FE <-> 1x BE. The possibilities for open/close expand to:
```
open 1st FE
open 2nd FE
close 2nd FE
close 1st FE

open 1st FE
open 2nd FE
close 1st FE
close 2nd FE

open 2nd FE
open 1st FE
close 2nd FE
close 1st FE

open 2nd FE
open 1st FE
close 1st FE
close 2nd FE
```
Conclusion? The framework has to be able to deal with cases where FE that brought given BE to life, is closed _before_ any subsequent FE that were opened on the same BE. It effectively _reparents_ the BE to the next FE in line which is still operational. When that happens, `be_substream->runtime` pointer is **reassigned**. And with it, information previously found in `be_substream->runtime->private_data` is no longer accessible for that BE.

Now, HDAudio implementation located in `sound/pci/hda` works with `substream->runtime->private_data` containing a valid pointer to `struct hdac_stream` in mind. Combine that with reparenting and we get guaranteed panic e.g.: in `hdmi_pcm_close()` when closing one of FE connected to single BE for a HDMI endpoint.

Reply:
we've removed the use of private data in SOF for the BE. see https://github.com/thesofproject/linux/pull/3908

Reply:
You would have to elaborate as that does not seem true. hda-pcm.c is still full of it. More over, if you actually drop `substream->runtime->private_data` while still utilizing patches in `sound/pci/hda/patch*.c` you will get panics even if no reparenting gets involved.

Reply:
sound/soc/sof/intel/hda-pcm.c is ONLY used for the FE.
sound/soc/sof/intel/hda-dai.c is ONLY used for the BE.
the BE does not rely on the private information any longer - are we in agreement on this?